### ISSUE TITLE
retrier: RunCtx is not immediately cancellable

### ISSUE THREAD

**Issue by rjeczalik** [eapache/go-resiliency#41]
When `ctx` passed to `RunCtx(ctx, func(ctx ...` is cancelled, the RunCtx method does not return early, instead it waits until the current backoff time passes.

This is caused by the following code:

https://github.com/eapache/go-resiliency/blob/b912f08afda846a1202e193fa5890e9da1cc5baa/retrier/retrier.go#L98-L103

While researching the proper API usage I stumbled upon [a blog post](https://arangodb.com/2020/09/a-story-of-a-memory-leak-in-go-how-to-properly-use-time-after/) that proposes such construct.

In my opinion this blog post is a bit misleading, because it applies a solution for `time.After` to a `time.NewTimer` usage, which makes no sense.

So I used [sourcegraph.com/search]() to search for some of the usages of the `time.NewTimer` which will confirm my above conclusion.

So I found the following NewTimer usage in the `net/http` package of the standard library:

```go
	timer := time.NewTimer(nextPollInterval())
	defer timer.Stop()
	for {
		if srv.closeIdleConns() {
			return lnerr
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-timer.C:
			timer.Reset(nextPollInterval())
		}
	}
```
https://github.com/golang/go/blob/b2dbfbfc2315557815e1d5de12f28ed57f60958a/src/net/http/server.go#L3008-L3020

As shown above, there is no need to wait on `timer.C` when `ctx` is cancelled, we just need to ensure the timer is stopped when it goes out of scope.

So combining all the above, the issue I am reporting (that RunCtx does not return immadiately when ctx is cancel) could be fixed by the following patch:

```patch
diff --git a/retrier/retrier.go b/retrier/retrier.go
index e1b4349..9914a29 100644
--- a/retrier/retrier.go
+++ b/retrier/retrier.go
@@ -69,6 +69,7 @@ func (r *Retrier) RunCtx(ctx context.Context, work func(ctx context.Context) err
 // is returned to the caller regardless. The work function takes 2 args, the context and
 // the number of attempted retries.
 func (r *Retrier) RunFn(ctx context.Context, work func(ctx context.Context, retries int) error) error {
+       var timer *time.Timer
        retries := 0
        for {
                ret := work(ctx, retries)
@@ -81,7 +82,11 @@ func (r *Retrier) RunFn(ctx context.Context, work func(ctx context.Context, retr
                                return ret
                        }

-                       timer := time.NewTimer(r.calcSleep(retries))
+                       if d := r.calcSleep(retries); timer == nil {
+                               timer = time.NewTimer(d)
+                       } else {
+                               timer.Reset(d)
+                       }
                        if err := r.sleep(ctx, timer); err != nil {
                                return err
                        }
@@ -96,10 +101,7 @@ func (r *Retrier) sleep(ctx context.Context, timer *time.Timer) error {
        case <-timer.C:
                return nil
        case <-ctx.Done():
-               if !timer.Stop() {
-                       <-timer.C
-               }
-
+               timer.Stop()
                return ctx.Err()
        }
 }
```

@eapache If you agree with the above reasoning, would you want me to send the above as PR?

**Comment by eapache** [comment]
> I stumbled upon [a blog post](https://arangodb.com/2020/09/a-story-of-a-memory-leak-in-go-how-to-properly-use-time-after/) that proposes such construct.

More importantly, it's in [the golang docs on `Timer.Stop`](https://pkg.go.dev/time#Timer.Stop).

> In my opinion this blog post is a bit misleading, because it applies a solution for time.After to a time.NewTimer usage, which makes no sense.

The [golang docs](https://pkg.go.dev/time#After) explicitly state that `time.After` is equivalent to a call to `time.NewTimer`.

> When ctx passed to RunCtx(ctx, func(ctx ... is cancelled, the RunCtx method does not return early, instead it waits until the current backoff time passes.

In my tests, this is not the case. `timer.Stop()` does not wait for the timer, and `<-timer.C` only gets run if `timer.Stop()` fails, in which case there must already be a value ready to be read from the channel.

If you can provide a concrete test case to demonstrate this delay, please do so. Based on my review of the golang documentation, the code is already correct.

**Comment by rjeczalik** [comment]
> More importantly, it's in [the golang docs on Timer.Stop](https://pkg.go.dev/time#Timer.Stop).

The `if !timer.Stop() { <- timer.C }` is used to drain the timer, before reusing it. In the code I mentioned, this is not the case.

> In my tests, this is not the case. timer.Stop() does not wait for the timer, and <-timer.C only gets run if timer.Stop() fails, in which case there must already be a value ready to be read from the channel.

And here is the problem exactly - when `ctx` is cancelled the timer is not read, obviously, but we still wait on channel. My point is we do not need to wait (drain the timer) when `ctx` is cancelled, because we are going to throw away it regardless.

> If you can provide a concrete test case to demonstrate this delay, please do so. Based on my review of the golang documentation, the code is already correct.

Catching timing errors in a nice test-case is usually a bit tedious work, I can give you an reproducer instead. I will update this thread with the code.
